### ISSUE TITLE
[V2] Why QKV are transposed and Where is Q padded?

### ISSUE THREAD

**Issue by FdyCN** [Dao-AILab/flash-attention#430]
@tridao Could you please help me with these questions？

Q1: When I learned the cuda source codes. I found that Q\K\V will be transposed from `[batch, num_head, seq_len, head_dims]` to `[batch, seq_len, num_head, head_dims]`？Is there any reason cause i found pytorch doing this too. (Because when using kv-cache , seq_len_q = 1？)

comment in `mha_fwd` function suggests Q\K\V are transposed：
https://github.com/Dao-AILab/flash-attention/blob/d30f2e1cd50185c98ed88c0684b4a603f15bee37/csrc/flash_attn/flash_api.cpp#L187C41-L187C86

pytorch used v1: https://github.com/pytorch/pytorch/blob/f8817d8ac836204535f81e973e6fddc94ca1859f/aten/src/ATen/native/transformers/cuda/attention.cu#L669

Q2: Then I checked the cuda grids:
```
const int num_m_block = (params.seqlen_q + Kernel_traits::kBlockM - 1) / Kernel_traits::kBlockM;
dim3 grid(num_m_block, params.b, params.h);
```
I figure that you want a CTA to handle kBlockM rows on seqlen_q in Q，which is transposed before and seqlen_q is packed into batch_size？If it is correct that "a CTA to handle kBlockM rows on seqlen_q in Q", then, when using kv-cache，and seqlen_q = 1(after 1st loop)，we should pad seqlen_q to kBlockM or 16 for TensorCore process at least？but I don't find the code to pad Q？

i'm really confused why transpose cause I think it's not match the policy for grids allocation，and seqlen_q should be padded when seqlen_q = 1 or not？ If yes, where does it happen？

wondering for your reply, thank you so much for all your jobs.

**Comment by tridao** [comment]
Q1: the (batch, seqlen, nheads, headdim) format is more efficient. Q, K, V are output of a matmul with shape (batch, seqlen, hiddendim) with hiddendim = nheads * headdim, which are then reshaped to (batch, seqlen, nheads, headdim). This reshape does not incur any memory copy (purely just manipulating metadata like shapes and strides).

Q2: each CTA takes care of (kBlockM * seqlen_k) elements (i.e. kBlockM rows on seqlen_q in Q). If seqlen_q is less than kBlockM (e.g. seqlen_q = 1 and kBlockM=128) then we implicitly pad it with zero. This [line](https://github.com/Dao-AILab/flash-attention/blob/d30f2e1cd50185c98ed88c0684b4a603f15bee37/csrc/flash_attn/src/flash_fwd_kernel.h#L272) reads in Q with some guard (i.e. if some indices are out of bound for Q, we fill it with some other values instead of reading from global memory). There's no need to explicitly pad Q.

**Comment by FdyCN** [comment]
> Q1: the (batch, seqlen, nheads, headdim) format is more efficient. Q, K, V are output of a matmul with shape (batch, seqlen, hiddendim) with hiddendim = nheads * headdim, which are then reshaped to (batch, seqlen, nheads, headdim). This reshape does not incur any memory copy (purely just manipulating metadata like shapes and strides).
> 
> Q2: each CTA takes care of (kBlockM * seqlen_k) elements (i.e. kBlockM rows on seqlen_q in Q). If seqlen_q is less than kBlockM (e.g. seqlen_q = 1 and kBlockM=128) then we implicitly pad it with zero. This [line](https://github.com/Dao-AILab/flash-attention/blob/d30f2e1cd50185c98ed88c0684b4a603f15bee37/csrc/flash_attn/src/flash_fwd_kernel.h#L272) reads in Q with some guard (i.e. if some indices are out of bound for Q, we fill it with some other values instead of reading from global memory). There's no need to explicitly pad Q.

@tridao Thank you for your reply.

Q1: that really makes sense, to remove fusable transpose in MHA.

Q2: if kBlockM = 64\128 when seq_len = 1，there are many redundant calculation with padding 0。Is the performance still better than w/o TensorCore？ Or maybe try to use smaller kBlockM like 8\16 except the 1st loop will be better？(cause all loops except the 1st loop that seq_len_q = 1)

**Comment by tridao** [comment]
When seqlen_q is small, the bottleneck is just to load K and V as fast as possible. There's ongoing work to optimize for this case, which we hope to release soon (probably 1-3 weeks).

**Comment by lms-mt** [comment]
> Q1: the (batch, seqlen, nheads, headdim) format is more efficient. Q, K, V are output of a matmul with shape (batch, seqlen, hiddendim) with hiddendim = nheads * headdim, which are then reshaped to (batch, seqlen, nheads, headdim). This reshape does not incur any memory copy (purely just manipulating metadata like shapes and strides).
> 
> Q2: each CTA takes care of (kBlockM * seqlen_k) elements (i.e. kBlockM rows on seqlen_q in Q). If seqlen_q is less than kBlockM (e.g. seqlen_q = 1 and kBlockM=128) then we implicitly pad it with zero. This [line](https://github.com/Dao-AILab/flash-attention/blob/d30f2e1cd50185c98ed88c0684b4a603f15bee37/csrc/flash_attn/src/flash_fwd_kernel.h#L272) reads in Q with some guard (i.e. if some indices are out of bound for Q, we fill it with some other values instead of reading from global memory). There's no need to explicitly pad Q.

@FdyCN  As for Q1, in Torch, FlashAttention is adapted to the `torch.nn.functional.scaled_dot_product_attention`, which needs query to have shape [bs, head_num, q_seq_len, head_dim], same with K,V. So this really brings extra overhead in Torch.
